### ISSUE TITLE
Only show RenderWarnings if features are really missing

### ISSUE THREAD

**Issue by flying-sheep** [naturalcrit/homebrewery#850]
Users using a standards-abiding browser (such as Firefox) shouldn’t have to see the warning. Instead of testing if the user uses Chrome, the actual features should be tested. Instead of

https://github.com/naturalcrit/homebrewery/blob/f28ed3d52ebcf048244d137252b1efd9f951a9a2/shared/homebrewery/renderWarnings/renderWarnings.jsx#L24

we could [test](https://developer.mozilla.org/en-US/docs/Learn/Tools_and_testing/Cross_browser_testing/Feature_detection#CSS) if the browser supports all necessary features and only display the warning if it doesn’t.

**Comment by tenthirtyone** [comment]
And Firefox added support for the column issue and users can change it in about:config

**Comment by calculuschild** [comment]
I personally only test functionality with Chrome. This is probably a more correct way to do things, but while we are at it are there any other compatibility issues that we should warn users about that are still present in Firefox?

**Comment by flying-sheep** [comment]
If you’re using web standards everywhere and consult the compat tables before putting in something new, Firefox should generally just work. From my 10 years of web dev experience: Firefox is consistently better at implementing new stuff without bugs, while Chrome likes to implement stuff more quickly but often broken (then they take a long time to fix them if they ever do)

**Comment by calculuschild** [comment]
Right I see what you're saying, but there may be features we already use that Firefox and Chrome still do not implement identically, so let me rephrase my question:

Are there any remaining existing incompatibilities with the Homebrewery and Firefox? We consistently get bug reports and complaints on the subreddit about failed rendering, printing issues, etc. that often come down to use of a non-Chrome browser and the slight differences in feature implementation. For instance, as far as I am aware, printed background images and shadows do not appear as they should in the resulting PDF when printed from Firefox.

Given those existing issues, my current thought process is this:
We _could_ go through each and every incompatibility and make a custom warning for users, but that defeats the purpose of focusing on supporting one browser, that is, reducing the need to test on all browsers. Or we can just leave this general warning that says "hey, you can still use this site, but we noticed you aren't using Chrome so if things break... try Chrome first."

In light of this discussion, my initial reaction would be to leave the warning, but reword it to reflect this idea generally instead of listing this specific standard that Firefox is now up-to-date with: 

> **Built for Chrome**
> Other browsers may have unexpected behaviors.
> If you experience issues with your document not
> rendering or printing properly, please try using
> the latest version of Chrome before submitting
> a bug report.

Also, the warning should only show once per user unless they clear their cookies or move to a new machine. I know it's still a bit annoying but we've tried to make it as minimally intrusive as possible.

Of course, if anyone wishes to submit PRs to fix any incompatibility issues with other browsers they are more than welcome to do so. Fixing those types of issues just aren't a priority for me or @stolksdorf.

Thoughts?

**Comment by flying-sheep** [comment]
Well, my point is that if you rely on behaviors that are different in Chrome and Firefox, chances are you rely on some Chrome bug, not standard behavior. Of course that doesn’t always have to be the case: E.g. I’m pretty sure Chrome was at some point leaps and bounds ahead of Firefox in PDF printing, as you said.

**Comment by Falconloft** [comment]
This is more of an anecdotal argument. In most tests (including this one from this month: https://www.lifewire.com/firefox-quantum-vs-chrome-4176159), Chrome renders CSS more accurately. Most 'bugs' that Chrome runs into are caused by coding specifically for Firefox rather than sticking to standard CSS. I understand people wanting to use Firefox for security, but the accuracy issue really needs to be dropped.

**Comment by flying-sheep** [comment]
Funny, I had the exact opposite experience when I wrote stuff heavily relying on SVG, some 4 years ago. Chrome. Bugs. Everywhere.

Also without linking any results, your article is just claiming stuff, not proving stuff. Don’t get me wrong, this might very well be true, it just wasn’t in my practical experience: Anecdotal for sure, but is the number of pixel-perfect renderings important or occasional heavy breakage for something that should work according to spec?

**Comment by Falconloft** [comment]
SVG is an entirely different beast, however, aside from some specular lighting issues or stacking issues (if you let the renderer decide), it's pretty much uniform. You can sometimes see where a pixel was rounded to a different side of a line, but that's about it. The same code will look essentially identical in all modern browsers.

If you were seeing bugs in Chrome and not in Mozilla, it was more likely a result of code that wasn't clean, as Chrome will simply ignore incorrect code instead of trying to parse it in many instances. I don't know how Mozilla handles it.

**Comment by flying-sheep** [comment]
Ah, found my rant about chrome’s bugginess. Ran into 7 bugs (CSS, canvas, and SVG) in a single project, Chrome was (and maybe still is) really impressively broken: https://github.com/unconed/MathBox.js/issues/27#issuecomment-42128784

**Comment by calculuschild** [comment]
Even in your quoted comment, the developer mentions that for his own particular use case, Firefox was not performing appropriately (and from his demos, still isn't). Like him, our application seems to perform best in Chrome. There are always sure to be several features that Firefox follows more closely to standards as well as areas where it falls short. 

Whether it works well or not in Firefox is moot; what we mean to convey to the user is that this website _does_ work in Chrome, so try that if something looks broken. There is no desire (or time) on our part to rewrite large chunks of the styling and rendering code to make performance favor Firefox in exchange for possibly breaking things that are working fine on Chrome right now, especially when Chrome currently has almost 8x the market share of Firefox. As it stands, the intended purpose of the RenderWarning remains valid (if not the specific text of the warning): "From the start, Chrome was the only browser that had the features we needed. Even though other browsers may have fixed many of these issues, we can only be confident that things work in Chrome, so if you see a weird bug, be warned that we haven't tested this in Firefox and that may be the reason."

Again, if _you_ or anyone else would like to submit PRs to fix functionality for Firefox, go right ahead, and we can incorporate them. After all of this discussion, it seems (to me) that it's not worth the extra mess of cataloging each and every feature that we are using and running a massive validation script for everyone, because even testing the presence of a given function (as in #896) may not be enough if one browser implements it differently and the result shows graphical glitches or other errors. It's just not practical for this type of hobby project.

**Comment by flying-sheep** [comment]
Perform as in like frames per seconds, not accuracy, so that’s completely beside the point.

**Comment by calculuschild** [comment]
Closed for now by #937 .
